Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -16,16 +16,94 @@
 
 #include "Firestore/core/src/firebase/firestore/remote/serializer.h"
 
-// TODO(rsgowman): These are (currently!) unnecessary includes. Adding for now
-// to ensure we can find nanopb's generated header files.
-#include "Firestore/Protos/nanopb/google/firestore/v1beta1/document.pb.h"
-#include "Firestore/Protos/nanopb/google/protobuf/timestamp.pb.h"
+#include "external/nanopb/src/nanopb/pb_decode.h"
+#include "external/nanopb/src/nanopb/pb_encode.h"
 
 namespace firebase {
 namespace firestore {
 namespace remote {
 
-Serializer::Serializer() {
+using firebase::firestore::model::FieldValue;
+
+Serializer::TypedValue Serializer::EncodeFieldValue(
+    const FieldValue& field_value) {
+  Serializer::TypedValue proto_value{
+      field_value.type(), google_firestore_v1beta1_Value_init_default};
+  switch (field_value.type()) {
+    case FieldValue::Type::Null:
+      proto_value.value.null_value = google_protobuf_NullValue_NULL_VALUE;
+      break;
+    default:
+      // TODO(rsgowman): implement the other types
+      abort();
+  }
+  return proto_value;
+}
+
+void Serializer::EncodeTypedValue(const TypedValue& value,
+                                  std::vector<uint8_t>* out_bytes) {
+  bool status;
+  // TODO(rsgowman): how large should the output buffer be? Do some
+  // investigation to see if we can get nanopb to tell us how much space it's
+  // going to need.
+  uint8_t buf[1024];
+  pb_ostream_t stream = pb_ostream_from_buffer(buf, sizeof(buf));
+  switch (value.type) {
+    case FieldValue::Type::Null:
+      status = pb_encode_tag(&stream, PB_WT_VARINT,
```

Comment:
Since nanopb requires us to get down to the level of encoding the tags for ourselves I'd like for us to build up a library of ways to concisely be able to describe this. Consider as an example, the code to compose leveldb keys (which operate at mostly the same granularity):

https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Local/FSTLevelDBKey.mm#L456

Since we have error handling to consider that isn't a consideration there, perhaps we should embed the buffer and status in a writer-type object (say e.g. `ProtoWriter`) to avoid littering this code as much with error checking.

Total straw man:
```c++
class ProtoWriter {
 public:
  EncodeTag(uint32_t tag) {
    if (!last_status) return;

    last_status = pb_encode_tag(&stream_, PB_WT_VARINT, tag);
  }

  void EncodeVarInt(int64_t value) {
    if (!last_status_) return;
    last_status_ = pb_encode_varint(&stream_, value);
  }

  // maybe
  EncodeTaggedVarInt(uint32_t tag, int64_t value) {
    EncodeTag(tag);
    EncodeVarInt(value);
  }

  util::Status AppendTo(std::vector<uint8_t>* out_bytes, const char* description) {
    if (!last_status) {
      // compose error status using description
      // abort() for now for lack of Status.
    }
    // other checking
    out_bytes->insert(out_bytes->end(), buf_, buf_ + stream_.bytes_written);
    return util::OkStatus();
  }

 private:
  bool last_status_;
  uint8_t buf_[1024];
  pb_ostream_t stream_;
};
```
And this code becomes:
```c++
ProtoWriter writer;
switch (value.type) {
  case FieldValue::Type::Null:
    writer.EncodeTag(google_firestore_v1beta1_Value_null_value_tag);
    writer.EncodeVarInt(value.value.null_value);
    return writer.Append(out_bytes, "Null value");
    break;
  }
}
```
In general I want us to abstract away the fiddly details of the proto wire encoding, weird buffering, and the raw nature of the nanopb API so that we can still understand and change this code within a few months.

With that said, if these patterns aren't yet emerging it may not yet be time for this kind of abstraction.

Reply:
Agree with all of this, but would like to defer it. (In fact, the next PR already has some of this implemented.)

Reply:
SGTM. This took shape somewhat organically in the leveldb case too until I finally broke down and did the grand refactoring. As long as this is top of mind we're good for now.